include/aws/common/byte_buf.h

```diff
@@ -451,6 +451,15 @@ int aws_byte_buf_append_and_update(struct aws_byte_buf *to, struct aws_byte_curs
 AWS_COMMON_API
 int aws_byte_buf_append_null_terminator(struct aws_byte_buf *buf);
 
+/**
+ * Expand the buffer appropriately to meet the requested capacity.
+ *
+ * If the the buffer's capacity is currently larger than the request capacity, the
+ * function does nothing (no shrink is performed).
+ */
+AWS_COMMON_API
+int aws_byte_buf_reserve_dynamic(struct aws_byte_buf *buffer, size_t requested_capacity);
```

Comment:
i actually run into same problem couple days ago and could have sworn we already had helper. So was thinking of implementing it. 

Minor suggestion on naming: maybe something like aws_byte_buf_ensure_capacity?